Updating `self` syntax and adding `static` member variables (#7016)
carbon-language/carbon-lang: proposals/p7016.md

-   [Alternatives considered](#alternatives-considered)
    -   [`class` modifier for non-instance members](#class-modifier-for-non-instance-members)
    -   [`shared` or `global` keywords for non-instance fields](#shared-or-global-keywords-for-non-instance-fields)
    -   [`static` for non-instance class functions as well as non-instance fields](#static-for-non-instance-class-functions-as-well-as-non-instance-fields)
    -   [Distinct `method` introducer](#distinct-method-introducer)
    -   [Special `self` syntax in square brackets `[]`](#special-self-syntax-in-square-brackets-)

<!-- tocstop -->

## Abstract

Update the syntax for class (and interface/`impl`) methods to move `self` into
the parameter parentheses `()` and make its type binding optional (defaulting to
`Self`). Introduce the `static` keyword for non-instance fields to indicate
static storage. Reflects the decision in leads issue
[#6931](https://github.com/carbon-language/carbon-lang/issues/6931).

## Problem

Carbon's syntax for defining instance methods and non-instance members (static
members in C++ terminology) needs refinement to balance ergonomics, clarity, and

> consistency across classes, interfaces, and `impl` blocks.

Reviewer: I find this paragraph very hard to get through as a reader, but I'm struggling to give effective feedback, because I'm not clear on what you intend the reader to take away from it (which is a big part of why I find it hard to get through).

Author: Sorry about that, I didn't proofread this section nearly enough. The paragraph I think ended up being nonsense.

The key I think is to make a point that we need to take the time to revisit and make a durable decision around this syntax given its importance, with a side problem of needing to add non-instance data member syntax.

I've tried to rewrite this section, but let me know if its still confusing.

The current syntax for method definitions can be heavy, and there is no clear
way to declare non-instance fields (class variables) that works uniformly across
all contexts (classes, interfaces, `impl`s).

## Background

Leads discussed various syntax tweaks for class members in issue
[#6931](https://github.com/carbon-language/carbon-lang/issues/6931). The

> discussions focused on find a design that scale from simple classes to complex

Reviewer: ```suggestion
discussions focused on finding a design that scales from simple classes to complex
```

> interface implementations.

Reviewer: For me this sentence raises more questions than it answers -- how does the syntax for `self`, or for non-instance data members, scale (or fail to scale) with the complexity of the enclosing context? This is particularly important because placing this sentence in the "Background" section (especially when that section is so short) tells me that I need to understand it in order to understand the rest of the proposal.

Author: Sorry, a confusing sentence. Updated to just talk about the fact that this problem as discussed extensively in the issue.

> Key tension points included:

Reviewer: Does "tension points" here mean something more than "questions"?

Author: No, rephrased to just call these questions.

-   Where to place `self` (in `[]` versus `()`).
-   How to distinguish instance methods from non-instance functions.
-   How to declare non-instance fields without overloading keywords or creating

>     confusion in interfaces vs classes.

Reviewer: This seems more like part of the problem statement than part of the background.

Author: Agreed, moved.

## Proposal

We propose the following updates to the syntax for class members:

1.  **Move `self` to the `()` parameter list**: Instance methods declare `self`

>     as the first parameter in the standard parameter parentheses. This aligns

Reviewer: The term "standard parameter parentheses" seems to be new. Up to now we've been calling this the "explicit parameter list"; why the terminology shift?

Author: Wasn't intending a new term, just trying to be descriptive. But happy to use the current terminology of "explicit parameter list".

>     methods with the **partial-application model** (see Rationale).

Reviewer: Here again, as a reader I don't understand what I'm supposed to take away from this sentence, particularly if I'm reading this document in the order it's written (where this is the first mention of the partial-application model).

Author: Reworded to more directly explain that this is making method calls more of syntactic sugar for function calls that passes the object as the first argument. Does that work better?

2.  **Make `self` type optional**: Omit `: Self` to default to the enclosing

>     type (for example, `fn F(self)`).

Reviewer: It's surprising to describe this in terms of omitting `: Self`, but then not mention `Self` when describing what it defaults to. Is "the enclosing type" potentially different from `Self`?

Author: Sure, reworded to be more clear.

3.  **Use `static` for non-instance fields**: Use `static var` for data members

>     tied to the type rather than instances.

Reviewer: ```suggestion
    that are part of the type rather than part of instances of the type.
```

("Tied to" is vague)

4.  **No keyword for non-instance methods**: Functions without a `self`
    parameter are automatically non-instance functions.

> As a consequence of this model, methods are fundamentally functions and can be

Reviewer: What model? This phrasing seems like it's intended to come after a discussion of the partial-application model, but the next sentence wouldn't make sense in that context, so I'm not sure what to suggest.

Author: Reworded a bit to try to more directly state in brief summary the consequence of the proposal (which was my intent here).

called using standard function call syntax (for example, `MyClass.Method(obj)`),
not just method call syntax (`obj.Method()`). This makes method calls
essentially perform partial application.

> However, we retain the current model for the meaning of `obj.Method` -- it is a

Reviewer: This compounds my confusion about the model alluded to in the previous paragraph -- what contrast is "however" pointing to? What other model for `obj.Method` are we disclaiming here?

Author: Reworded to try to be more clear as above.

"bound member function" that adapts `obj`. This is important as we don't know
how to convert `obj` to the `self` parameter of `Method` until we actually try
to call it with an argument list. Similarly, we expect
`obj.(Class.Method)(args)` to follow the same model, and `obj.(Class.Method)` to

> form a callable bound method adapter. For a given constant expression of a

Reviewer: Is a "bound method adapter" different from a "bound member function"? If not, why the change in terminology?

Author: The goal was just to clarify that there is still an adaptor of the object. Hopefully reworded to make that more clear?

> method `M`, we expect `x.(M)` means `x as BoundMethod(TypeOf(x), M)`, and given

Reviewer: What does it mean for a constant expression to be "of a method"?

Author: Reworded a bit, but the key is that this is a constant expression referring to a method?

> `b: BoundMethod(T, M)`, `b(args)` means `M(x as T, args)` where `x` is the bound

Reviewer: Where did `BoundMethod` come from? You're presenting it as the status quo, but I can't find it anywhere in our docs.

Author: Sorry, this was supposed to be illustrative of the instance binding model, and just trying to say that the same model applies.

Instead, I've said directly the above and linked to the model. That probably works better.

object.

## Details

### Instance methods

Instance methods declare `self` as the first parameter in the `()` list. The

> type binding `: Self` is optional and defaults to `: Self`.

Reviewer: The term "type binding" seems to be new. Is that intentional?

Author: No, fixed the wording here and elsewhere I found it.

> type binding `: Self` is optional and defaults to `: Self`.

Reviewer: This PR should probably update `pattern-matching.md`, since it's introducing a new pattern syntax.

Author: Done.

```carbon
class Point {
  var x: i32;
  var y: i32;

  // Explicit type binding.
  fn GetX(self: Point) -> i32 { return self.x; }

  // Omitted type binding (defaults to Point).
  fn GetY(self) -> i32 { return self.y; }

  // By-reference binding.
  fn SetX(ref self, new_x: i32) { self.x = new_x; }
}
```

#### Calling without method syntax

Because methods are modeled as functions with an explicit `self` parameter, they

> can be called directly by way of the type name without using dot-notation:

Reviewer: `Point.GetX(p)` sure looks to me like it's using dot-notation. Maybe say instead something like "they can be called as ordinary functions using their qualified names, with the `self` argument as the first element of the explicit parameter list".

Author: Done.

```carbon
fn F(p: Point) {
  // Standard method call syntax:
  var x1: i32 = p.GetX();

>   // Non-method call syntax (partial-application style):

Reviewer: How is this "partial-application style"? IIUC this syntax (unlike the method-call syntax) cannot perform partial application -- you have to pass all the arguments at once.

Author: Removed.

}
```

### Class functions and static fields

Non-instance class functions don't use any special syntax. However, non-instance
fields, now called _static fields_, use the `static` keyword to distinguish them
from instance fields.

```carbon
class Widget {
  // Non-instance field.
  static var count: i32;

  // Non-instance function.
  fn ResetCount() { count = 0; }
}
```

This syntax applies uniformly to classes, interfaces, and `impl` blocks.

## Rationale

This proposal effectively advances Carbon's goals by focusing on:

-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write):
    Moving `self` to the "normal" parameter list better matches the semantics of
    the object argument and the partial-application model. Making the `: Self`
    part optional reduces clutter in the most common cases and matches the
    syntax used in other languages like Rust.
-   [Software and language evolution](/docs/project/goals.md#software-and-language-evolution):
    Focusing `static` on storage (fields) provides a crisp definition that works
    unambiguously across classes, interfaces, and `impl` blocks.

> ### Unification and Partial-Application Model

Reviewer: I'm not sure whether to parse this as "(Unification and Partial-Application) Model" or "Unification and (Partial-Application Model)". In context, neither of them seems to fit.

Author: Tried a different phrasing entirely.

Moving `self` into the parentheses reinforces the conceptual model that method
calls are sugar for function calls with an explicit object argument.
Specifically, `obj.Method(args...)` is equivalent to
`Type.Method(obj, args...)`.

This **partial-application model** unifies methods and functions: there is only
one kind of function definition. A function is an "instance method" if and only
if its first parameter is named `self`.

> This avoids alternative more complex models such as:

Reviewer: These seem like they should be in "Alternatives Considered"?

Author: Agreed, removed here and incorporatingb them below where they weren't already.

-   **Receiver Model**: Where the object is a "magical" entity outside the type

>     system or parameter list (for example, C++ `this`).

Reviewer: How is `this` "outside the type system"? It has a type, and even arguably a declaration.

Author: Sorry, this didn't make much sense. removed all of this.

-   **Bound-Member Model**: Where `obj.Method` would form a "closure" or "fat
    pointer" object. Carbon avoids the complexity of bound-member types by

>     treating method access as a direct function call with the object applied.

Reviewer: This passage seems to be saying that in Carbon, `obj.Method` can only be written as part of a function call, and isn't a valid expression on its own. In other words, we're disallowing partial application. Is that right? If so, I think that needs to be stated earlier and much more explicitly, because the term "partial-application model" really invites the opposite understanding.

Author: I've removed this, agree that it was confusing.

This unification simplifies generic programming and C++ interoperability, as it
minimizes special exceptions for member functions.

The decision to use `static` for fields and _no keyword_ for non-instance
methods leans on the distinction between storage and behavior. `static` implies

> storage lifetime, which fits data fields well. For functions, the absence of a

Reviewer: The claim that "`static` implies storage lifetime" seems hard to reconcile with how that keyword is actually used in e.g. C++ and Java.

Author: Discussed this a bit live in person and updated these paragraphs to do a better job.

`self` parameter is sufficient to distinguish them from methods without
stretching the meaning of `static`. This distinction aligns with common
definitions of [static variables](https://en.wikipedia.org/wiki/Static_variable)

> focusing on lifetime contrasted with scope.

Reviewer: If `static` is all about storage duration and not about scope or instance binding, why do we not require it (or even allow it) on package-scope variables?

Author: My argument is that the _default_ storage for namespace/package variables _is_ static storage, but this is another place where we can have a good default and omit it, and currently my leaning is towards having only one way to write it rather than making it optional. Added this explicitly.

Reviewer: On reflection, I don't think we allow package- or namespace-scope variables at all (unless there was a proposal for them, and we just haven't updated places like [this](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/README.md#global-constants-and-variables)), so that was my mistake. If you want to preemptively settle the question of what their syntax would be, the phrasing should probably be more hypothetical; I've incorporated that into my suggestion in another comment.

Author: Given that its a negative thing, left it in the alternatives considered FWIW, but it does use the future-oriented wording now I think.

## Alternatives considered

### `class` modifier for non-instance members

> Instead of `static var`, using `class var` or `class fn` was considered.

Reviewer: ```suggestion
We considered using `class var` and `class fn` to mark non-instance methods and variables.
```

- The passive voice is awkward.
- The phrasing made it sound like we considered using `class fn` as an alternative to `static var`.

> **Rejection rationale:** `class` is already an introducer in Carbon. Using it as

Reviewer: This "**Rejection rationale:**" idiom is new, and seems unnecessarily boilerplatey -- I think every instance could be replaced with "However,"

a modifier adds an overloaded meaning. This meaning would also diverge from
usage in other languages. For example, Swift uses `class` versus `static` to
distinguish between class and struct methods, and has complex virtual dispatch
rules that aren't part of the Carbon design. These members also appear in
interfaces and `impl` blocks where the `class` keyword would be mildly
surprising.

### `shared` or `global` keywords for non-instance fields

Alternatives to `static` like `shared var` or `global var` were considered to

> avoid C++'s overloaded usage of `static`. **Rejection rationale:**

Reviewer: Here again, awkward use of passive when we can just say "we". Also, it's not clear why these, but not the previous alternative, were "to avoid C++'s overloaded usage of `static`".

Author: I think fixed.

    example these variables are very likely to be shared between threads.

    On the other hand, there would be interesting cases where we would do
    something _other_ than mark something as `shared`, but still end up sharing
    it between threads and potentially referring to its place set as `shared`.

    Overall, the risk of confusion outweighed the symmetry benefits where the
    meanings matched.

-   **`global`**: Most widely used to refer to _scope_ (global scope), whereas
    these members are explicitly scoped to a type, or potentially in the future
    an interface. See for example the start of the two Wikipedia articles on
    [Static variable](https://en.wikipedia.org/wiki/Static_variable) vs
    [Global variable](https://en.wikipedia.org/wiki/Global_variable). The most
    common distinction is that of "global" implicating _scope_ or _access_,
    while "static" more consistently means the _storage_ which is how we are
    using it.

### `static` for non-instance class functions as well as non-instance fields

Applying `static` to both `static var x: i32;` and `static fn F();` for
consistency with C++, C#, and Swift (see

> [Type Methods](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/methods#Type-Methods)).

Reviewer: This isn't a complete sentence.

Author: Fixed I think.

one kind of function definition. A function is an "instance method" if and only
if its first parameter is named `self`.

This avoids alternative more complex models such as:

-   **Receiver Model**: Where the object is a "magical" entity outside the type
    system or parameter list (for example, C++ `this`).
-   **Bound-Member Model**: Where `obj.Method` would form a "closure" or "fat
    pointer" object. Carbon avoids the complexity of bound-member types by
    treating method access as a direct function call with the object applied.

This unification simplifies generic programming and C++ interoperability, as it
minimizes special exceptions for member functions.

The decision to use `static` for fields and _no keyword_ for non-instance
methods leans on the distinction between storage and behavior. `static` implies
storage lifetime, which fits data fields well. For functions, the absence of a
`self` parameter is sufficient to distinguish them from methods without
stretching the meaning of `static`. This distinction aligns with common
definitions of [static variables](https://en.wikipedia.org/wiki/Static_variable)

> focusing on lifetime contrasted with scope.

Reviewer: This paragraph seems out of place; it doesn't have much to do with the topic of this section, but it has a lot to do with the topic of the "`static` for non-instance class functions as well as non-instance fields" section below.

Author: Yeah, reorganized a bunch to make the structure better I hope.

> ## Alternatives considered

Reviewer: The rationale above and the "Special `self` syntax in square brackets `[]`" section below both discuss the alternative of treating `self` differently from all other parameters, and the latter even presents the choice not to do that as "fundamental". That being the case, why not present it as an alternative considered?

Author: I believe I've adde the relevant alternatives now, although structured a bit differently.

### `class` modifier for non-instance members

Instead of `static var`, using `class var` or `class fn` was considered.
**Rejection rationale:** `class` is already an introducer in Carbon. Using it as
a modifier adds an overloaded meaning. This meaning would also diverge from
usage in other languages. For example, Swift uses `class` versus `static` to
distinguish between class and struct methods, and has complex virtual dispatch
rules that aren't part of the Carbon design. These members also appear in
interfaces and `impl` blocks where the `class` keyword would be mildly
surprising.

### `shared` or `global` keywords for non-instance fields

Alternatives to `static` like `shared var` or `global var` were considered to
avoid C++'s overloaded usage of `static`. **Rejection rationale:**

-   **`shared`**: This term is heavily associated with concurrency and thread
    safety (for example, shared ownership, shared pointers). Reserving it for
    class membership would collide with its use for future concurrency features
    (see
    [concurrency control design](https://docs.google.com/document/d/1WVWcmJdVBlapza_kPj2l3mOO-yw_hNXpb2u-Ren-I5M/)
    and

>     [shared memory in CUDA](https://developer.nvidia.com/blog/using-shared-memory-cuda-cc/)).

Reviewer 2: `shared` doesn't clash with CUDA. In [my original post](https://github.com/carbon-language/carbon-lang/issues/6931#issuecomment-4146359301) proposing `shared` I actually called out support for CUDA shared memory as an argument *in favour* of `shared`:
> for more exotic platforms like GPUs, we could share variables within a thread block/thread group/workgroup, which is a [common need](https://developer.nvidia.com/blog/using-shared-memory-cuda-cc/): `shared(workgroup) var cache: SomeKindOfCache;`

Hence I am confused that CUDA is being used here as an argument *against* `shared`.

Author: I don't really follow ...

My point is that CUDA already has a meaning of "shared", and I don't think the meaning we would want for these would _exactly_ match its meaning of shared.

Maybe this collision or overlap is something you see as a positive? If so, then I think I disagree.

    Some of these might be related and even highly compatible overlap, and in
    common cases it would provide a certain amount of semantic connection -- for
    example these variables are very likely to be shared between threads.

    On the other hand, there would be interesting cases where we would do
    something _other_ than mark something as `shared`, but still end up sharing
    it between threads and potentially referring to its place set as `shared`.

    Overall, the risk of confusion outweighed the symmetry benefits where the
    meanings matched.

-   **`global`**: Most widely used to refer to _scope_ (global scope), whereas
    these members are explicitly scoped to a type, or potentially in the future
    an interface. See for example the start of the two Wikipedia articles on
    [Static variable](https://en.wikipedia.org/wiki/Static_variable) vs
    [Global variable](https://en.wikipedia.org/wiki/Global_variable). The most
    common distinction is that of "global" implicating _scope_ or _access_,
    while "static" more consistently means the _storage_ which is how we are
    using it.

> 

Reviewer: Should these just be two separate alternatives?

Author: Switched to headings instead of bullets?

### `static` for non-instance class functions as well as non-instance fields

Applying `static` to both `static var x: i32;` and `static fn F();` for
consistency with C++, C#, and Swift (see
[Type Methods](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/methods#Type-Methods)).
**Rejection rationale:** The leads preferred to tie `static` strictly to
**storage duration** (lifetime). Non-instance data members have storage and a
storage duration of the entire program execution, which fits the definition of a
[static variable](https://en.wikipedia.org/wiki/Static_variable). Functions do
not have "storage", and so using the keyword for them would weaken this
interpretation.

### Distinct `method` introducer

> Using `method F()` instead of `fn F(self)`. **Rejection rationale:** The leads

Reviewer: Another incomplete sentence.

Author: Done.

currently prefer a uniform `fn` introducer for all functions, relying on the
presence of the `self` parameter to identify methods. This provides greater
consistency across all different kinds of functions, including even lambdas.

> ### Special `self` syntax in square brackets `[]`

Reviewer: What's "special" about this syntax?

Author: Fixed the heading.

Placing `self` in square brackets (for example, `fn F[self: Self](...)`) was
considered as it visually distinguished the object parameter from other

> parameters and has been the historical syntax in Carbon.

Reviewer: It seems like you're going out of your way to avoid using the term "implicit parameter list", or even mention that the square brackets might contain other things besides the `self` parameter. Is that intentional?

Author: No, adjusted to be just call them iplicit parameters and to directly mention that we might have runtime implicit parameters.

**Rejection rationale:** We expect users to strongly associate square brackets

> with compile-time parameters, as those are the typically "deduced" parameters to

Reviewer: It's not clear why "deduced" is being brought up here, especially in scare-quotes.

Author: No idea, reworded all of this.

functions. While we may add implicit runtime parameters in the future, even then
they would be significantly more rare than compile-time parameters. Using `self`

> in the same context loses that consistency. For example, if we pursue

Reviewer 2: I draw the opposite conclusion. Putting `self` in the deduced parameter list would *help* users learn what `[]` means, rather than lead them awry. Notably, `[]` in Python types and functions means "type parameter list" / "compile-time parameter list", so there is a major risk that people coming to Carbon misunderstand `[]`. The sooner we show that runtime variables can appear inside `[]`, the better IMO.

That's not to say that `self` should remain in the square brackets. I'm just saying that the argument being made in this section doesn't make sense to me.

Author: This section has been reworded substantially to hopefully be a bit more clear.

Still, it may not fully make sense to you. I think the key thing is if you feel there are important points of consideration that aren't covered at all -- then we should add them. I've added the one you mention here.

context-based distinguishing of compile-time versus runtime parameters as
suggested in
[issue #6932](https://github.com/carbon-language/carbon-lang/issues/6932), then
`self` would have to trigger an exception to the contextual rule.

There was also a general appreciation for the consistency of putting `self` in
the `()`s because there are many different ways in which its behavior matches
what we expect for "normal" parameters:

-   The object argument may be implicitly converted to match the type declared

>     for the `self` parameter, just like any other arguments.

Reviewer: But that applies just as much to the square brackets as the round ones, doesn't it?

Author: Tried to clean this up a bit to highlight the real distinction here -- the issue is more that you can do deduction _from_ the `self` argument because it is an explicit argument. That doesn't trivially generalize to implicit arguments? Hopefully the text is more clear here.

-   Lifetime and other semantics match that of arguments as they are tied to the
    call expression.

Fundamentally, the leads see two workable models for how methods work in Carbon
-- one of partial application of the object argument, and one of a truly special
and bespoke kind of parameter. The semantics we currently have match the first
of these well, and so it seems natural to embrace that model rather than having

> `self` in a different position.

Reviewer: This paragraph doesn't seem particularly related to the "Special `self` syntax in square brackets `[]`" alternative. See also my earlier question about why special treatment for `self` isn't an alternative-considered in its own right.

Author: Removed / reworded.

    -   [`class` modifier for non-instance member variables and functions](#class-modifier-for-non-instance-member-variables-and-functions)
    -   [Alternative keywords for non-instance member variables](#alternative-keywords-for-non-instance-member-variables)
        -   [`shared`](#shared)
        -   [`global`](#global)
    -   [`static` for non-instance member functions](#static-for-non-instance-member-functions)
    -   [`static` for package- and namespace-scope variables](#static-for-package--and-namespace-scope-variables)
    -   [Distinct `method` introducer](#distinct-method-introducer)

<!-- tocstop -->

## Abstract

Update the syntax for class (and interface/`impl`) methods to move `self` into
the parameter parentheses `()` and make the type in its binding optional
(defaulting to `Self`). Introduce the `static` keyword for non-instance member
variables to indicate static storage. Reflects the decision in leads issue
[#6931](https://github.com/carbon-language/carbon-lang/issues/6931).

## Problem

The placement of `self` in the parameter list has been a repeated source of
debate around the syntax design of Carbon -- grouping it with deduced parameters

> like types is surprising and nearly unprecedented in modern programming

Reviewer 3: ```suggestion
like types is surprising for some readers and nearly unprecedented in modern programming
```
(Just to make it clearer that this is subjective.) Or s/some/many/ if you prefer to say something stronger :)

`self`. It can always be called directly with the first argument passed to the
`self` parameter. When `Method` names a function with a `self` parameter, the
syntax `obj.Method` is just syntactic sugar for partially applying the `obj` to
the `self` parameter, which can then be called as a function by passing
arguments for the remaining parameters.

For non-instance member functions, the absence of a `self` parameter is
sufficient to distinguish them from methods.

#### Calling without method syntax

Because methods are modeled as functions with an explicit `self` parameter, they
can be called as ordinary functions using their qualified names, with the `self`
argument as the first element of the explicit parameter list:

```carbon
fn F(p: Point) {
  // Standard method call syntax:
  var x1: i32 = p.GetX();

  // Called as an ordinary function:
  var x2: i32 = Point.GetX(p);
}
```

## Rationale

This proposal effectively advances Carbon's goals by focusing on:

-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write):
    Moving `self` to the "normal" parameter list better matches the semantics of

>     the object argument and the partial-application model. Making the `: Self`

Reviewer 3: I'm not sure I accept this as rationale, since we picked the partial application model to match the syntax we chose, not the other way around :)

I think we should call out that putting the `self` in the parens matches the syntax of other explicit-self languages, like C++23, Python, and Rust, and reduces the clutter of having two sets of brackets in the common case. (Not coincidentally, much the same benefits as those you point out for making the `: Self` optional.)

Author: Yeah, sorry, this didn't get updated when trying to more clearly articulate the rationale as being rooted in aligning with other languages' use of `self`. Hopefully updated.

-   How to declare non-instance fields without overloading keywords or creating
    confusion in interfaces vs classes.

## Background

Leads discussed various syntax options for `self` placement and non-instance
members in issue
[#6931](https://github.com/carbon-language/carbon-lang/issues/6931).

## Proposal

We propose the following updates to the syntax for class members:

1.  **Move `self` to the `()` parameter list**: Instance methods declare `self`
    as the first parameter in the explicit parameter list. This aligns methods
    with a model where method calls are conceptually sugar for function calls
    with an explicit object argument.
2.  **Make `self` type optional**: Optionally allow omitting the type of `self`,
    which results in the type being exactly `Self`. Example:
    `fn MyMethod(self)`.
3.  **Use `static` for non-instance data members**: Use `static var` for data
    members that are part of the type rather than part of instances of the type.
4.  **No keyword for non-instance member functions**: Functions without a `self`
    parameter declared within a `class` are considered non-instance member
    functions without any extra syntax.

> Treating methods as functions with an explicit first parameter means they can be

Reviewer: ```suggestion
Treating methods as functions with an explicit `self` parameter means they can be
```

called using standard function call syntax (for example, `MyClass.Method(obj)`),
not just method call syntax (`obj.Method()`).

While methods can be invoked as regular functions, method syntax (`obj.Method`)
continues to form a "bound member function" that adapts `obj`. This allows the
call to be resolved appropriately even when `obj` requires conversion to match
the type of `self`. Similarly, we expect `obj.(Class.Method)(args)` to follow
the same model. Fundamentally, our goal is that this does not change the
[instance binding model](/docs/design/expressions/member_access.md#instance-binding).

## Details

### Member variables

_Member variables_ are any `var`s declared within a class context (or in the
future, potentially an interface or `impl` context).

Instance member variables, or _fields_, are unchanged in the current design.

This proposal adds support for non-instance member variables, or _static member
variables_, using the `static` modifier keyword on the `var` declaration. Static
member variables are associated with the type itself rather than instances of
the type, and have static storage duration.

```carbon
class Widget {

>   // Non-instance field.

Reviewer: This usage is inconsistent with the definition of "field" above. I think your definition makes the right tradeoff: we should reserve the short name "field" for instance data members, because they're the overwhelmingly more common case. I like your previous suggestions on Discord, to call these "class variables".

There are a few other uses of "fields" below that refer to class variables; those should be updated as well.

Author: Yeah, Gemini really wanted to call these "non-instance fields", and I didn't catch a bunch of the places that needed to move away from that. Tried to much more systematicaly fix this.

Author: Wow, it was literally in teh proposal title. Sorry about that.

  static var count: i32;

  // Non-instance function.
  fn ResetCount() { count = 0; }
}
```

This syntax should apply uniformly to classes, and in the future if we add

> associated fields, interfaces, and `impl` blocks.

Reviewer: This is confusing. We _have_ added interfaces and `impl` blocks, we just haven't added static variables to them, and I can't tell what "associated fields" are meant to be, particularly if they're distinct from static variables in interfaces and `impl` blocks. The fact that package/namespace scope variables are exceptions should probably also be mentioned here. How about something like this?

```suggestion
This syntax won't apply to package- and namespace-scope variables (if we
support them) because static storage duration is the default in those contexts.
It is currently limited to classes, but if we allow such variables in other contexts
that don't have that clear default, like interfaces and `impl` declarations, they
should use the same syntax.
```

Reviewer: It looks like you moved this paragraph to the "`static` for package- and namespace-scope variables" alternative-considered, which is now the _only_ section that documents the decisions to not use `static` in package and namespace scopes, and to potentially use it in interface and `impl` scopes. I think that'll be confusing -- normally the Proposal and Details sections tell you _what_ is proposed, and then the Rationale and Alternatives tell you _why_ it's proposed.

I suggest keeping this paragraph here (in some shape or form).

Author: Done and tried to get the wording closer to your suggestion.

The decision to use `static` for fields because it implies storage lifetime,
which fits data fields well. This implication comes both from the origins of the
term (static allocation of storage versus dynamic allocation), and aligns
precisely with how existing languages use `static` on class members, especially
C++. This meaning also aligns with common definitions of
[static variables](https://en.wikipedia.org/wiki/Static_variable) focusing on

> lifetime contrasted with scope.

Reviewer: This is still quite hard for me to follow, even now that you've clarified your intent to me offline. I think a big part of the problem is that it's not clear what alternative this is being contrasted with -- presumably it should be "putting a different keyword here", but sometimes it seems to be contrasting with "putting `static` somewhere else".

This whole discussion seems like it belongs in the "alternatives considered" section anyway (probably as part of "Alternative keywords for non-instance member variables"), and re-framing it that way will probably help make it clearer.

Author: Sure, tried all of that.

> The storage of package- and namespace-scope variables should also be considered

Reviewer: This paragraph, also, should probably be an alternative considered, rather than part of the design details.

Author: Done.

to be "static" in the same way, and the omission of the keyword is because that

> storage being a strong default for package- and namespace-scope variables. The

Reviewer: Grammar nit

```suggestion
storage is a strong default for package- and namespace-scope variables. The
```

strong default for class members is instance- or field-storage, and so a keyword

> is needed to opt out of that default.

Reviewer: I believe we currently don't allow package- or namespace-scope variables (see e.g. [here](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/README.md#global-constants-and-variables)). Also, this is phrased as if the reader already knows that we've decided to omit the `static` keyword in these contexts, but I don't think you've actually said that.

```

Instance member functions, or _methods_, declare `self` as the first parameter

> in the `()` list. Specifying the type of `self` with `: Self` is optional and

Reviewer: ```suggestion
in the `()` list. Specifying the type of `self` is optional and
```

}
```

Moving `self` into the parentheses reinforces the conceptual model that method
calls are sugar for function calls with an explicit object argument.
Specifically, `obj.Method(args...)` is equivalent to
`Type.Method(obj, args...)`.

This adopts a **partial-application model** which unifies methods and functions.
A function is an "instance method" if and only if its first parameter is named
`self`. It can always be called directly with the first argument passed to the
`self` parameter. The method syntax of `obj.Method`, when `Method` names a
function with a `self` parameter, is just syntactic sugar for partially applying
the `obj` to the `self` parameter and calling the resulting function. This
includes the expression `obj.Method` itself being valid and being a callable

> expression.

Reviewer: ```suggestion
`self` parameter. When `Method` names a function with a `self` parameter,
the syntax `obj.Method` is just syntactic sugar for partially applying the `obj`
to the `self` parameter, which can then be called as a function by passing
arguments for the remaining parameters.
```

`obj.Method` isn't really "method syntax", because it applies to fields as well, and its desugaring does not include "calling the resulting function".

For non-instance member functions, the absence of a `self` parameter is
sufficient to distinguish them from methods. The decision is to not reuse
`static` here even though C++, Java, and many other languages do because this

> avoids stretching or diluting the meaning of `static` in Carbon. It let's us

Reviewer: Nit:
```suggestion
avoids stretching or diluting the meaning of `static` in Carbon. It lets us
```

    Focusing `static` on storage (fields) provides a crisp definition that works
    unambiguously across classes, interfaces, and `impl` blocks.

## Alternatives considered

> ### Treat `self` differently from _all_ parameters

Reviewer: ```suggestion
### Don't put `self` in either parameter list
```

We considered two approaches to modeling `self` that separated it from
parameters of any kind, implicit or explicit.

1.  We could place `self` in an independent position with a different set of
    delimiters such as:

        ```carbon
        class Point {
          var x: i32;
          var y: i32;

          // `self` as a separate component, likely in between the implicit and
          // explicit parameter lists.
          fn Create[T:! type]<self>(x: T) -> Self {
            return {.x = x as i32, .y = 0};
          }
        }
        ```

2.  We could make `self` implicit, similar to C++.

We have generally been happy with `self` being explicit instead of implicit and
so to an extent we didn't deeply consider (2) as that wasn't part of the problem

> we set out to solve.

Reviewer: It might help future readers to provide more of a breadcrumb trail here. For example, we could say that the rationale [here](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0722.md#full-receiver-type) still applies: we think it's important to be able to specify the type (and now the form) of `self`, even though we no longer require it.

Author: Done.

We did consider (1) but were unable to find a syntax that felt compelling.
Taking a new balanced delimiter is an especially difficult and expensive choice
for the syntax. And without that, most syntaxes felt verbose for something that
we expect to be extremely common and a natural "default".

### `self` syntax in the deduced parameter list `[]`

We considered placing `self` in the deduced parameter list (for example,
`fn F[self: Self](...)`), as this has been the historical syntax in Carbon.

However, users strongly associate the implicit parameter list in `[]`s with
deduced, compile-time parameters. While methods might deduce the _type_ of
`self`, the `self` parameter itself represents a _runtime_ argument passed by

> the caller. Placing it in the deduced parameter list conflates runtime arguments

Reviewer: I don't follow this sentence. Is it trying to contrast `self` with the type of `self`, or with other elements of the implicit parameter list? Is it contrasting them in terms of type vs. non-type, explicit vs. deduced, or compile-time vs. runtime?

I'd suggest instead folding the contrast into the previous sentence: "... with deduced, compile-time parameters, but `self` is a runtime parameter, and passed explicitly by the caller rather than being deduced". With that, I think you can also drop the next sentence about how it "conflates runtime arguments with compile-time parameters", which only addresses half of the problem (and "conflates" seems debatable to me).

Author: Done.

with compile-time parameters within that list. While we have syntactic
distinctions such as the `!` and the `self` keyword, the contextual collision
does still add some cognitive load for some readers. We do imagine potential
runtime implicit parameters in the future, but these are expected to be rare
enough to be easier to accommodate and to leverage distinguishing syntax.

In contrast to these challenges, putting `self` in the explicit parameter list
`()` aligns with the semantics of `self` as being an otherwise-normal parameter.
The type of it can be deduced, as it is an explicit parameter. And the lifetime
and passing semantics match that of arguments as they are tied to the call

> expression.

Reviewer: I recommend dropping these sentences. Neither of them contrasts with anything in the previous paragraph, and the problems they seem to be alluding to are either minor or outright non-existent.

Author: Done.

Last but not least, most other languages with explicit object parameters put
them at the start of the parameter list inside `()`s, see for example Python and

> Rust. Being consistent with how object parameters are modeled in other languages

Reviewer: Nit: this is a comma splice.

```suggestion
them at the start of the parameter list inside `()`s (see for example Python and
Rust). Being consistent with how object parameters are modeled in other languages
```

with its use for future concurrency features (see
[concurrency control design](https://docs.google.com/document/d/1WVWcmJdVBlapza_kPj2l3mOO-yw_hNXpb2u-Ren-I5M/)
and
[shared memory in CUDA](https://developer.nvidia.com/blog/using-shared-memory-cuda-cc/)).

The risk of confusion outweighed the benefits of identifying fields that are
conceptually shared across instances.

#### `global`

We considered using `global var` to mark non-instance member variables. However,
`global` is most widely used to refer to _scope_ (global scope), whereas these
members are explicitly scoped to a type (or potentially an interface). The most
common distinction is that of "global" implicating scope or access, while
"static" more consistently describes the storage model.

### `static` for non-instance member functions

We considered applying `static` to both member variables and functions (for
example, `static fn F()`), for consistency with C++, Java, C#, and Swift.
However, functions do not have instances or storage in the same way data members
do. The absence of a `self` parameter is sufficient to distinguish instance
methods from non-instance functions, making the keyword redundant for functions.

> There is a longer discussion about this in the detailed design as well.

Reviewer: I think that discussion should be moved here, because it's really about the rationale for not using `static` on functions, and not really about explaining the details of the design

Author: Done.

### Alternative keywords for non-instance member variables

The decision to use `static` at all in Carbon is relatively controversial. It
has a history in C++ of being used for a sprawling and seemingly ever-growing
set of things, in addition to the already subtle meanings inherited from C.

The decision to use `static` for member variables is specifically because it
implies storage lifetime, which fits data member variables well. This
implication comes both from the origins of the term (static allocation of
storage versus dynamic allocation), and aligns precisely with how existing
languages use `static` on members variables, especially C++. This meaning also
aligns with common definitions of
[static variables](https://en.wikipedia.org/wiki/Static_variable) focusing on

> lifetime contrasted with scope.

Reviewer: I'm afraid I'm still struggling with this paragraph. To begin with, "which fits data members well" seems like a non sequitur; it's really unclear how it relates to the rest of the sentence, or the rest of the paragraph. It gives the impression that we're trying to find the place where `static` fits best, instead of trying to find the keyword that fits this place best. If you want to frame the issue that way, I think you need to start by explicitly spelling out that we want `static` to have a single meaning in Carbon, and then build the rest of the paragraph to make the case that it fits better on variables than on e.g. functions. Otherwise, I suggest removing it.

The sentence about the "origins of the term" seems like an [etymological fallacy](https://en.wikipedia.org/wiki/Etymological_fallacy) -- even if "static" was first used in reference to storage, that history doesn't control how it's used now. I'd suggest striking that sentence, or moving it so that other, stronger arguments come first.

The last sentence seems like the strongest point, and I think you could make it even stronger: virtually all of Carbon's major peer languages (notably C++, Rust, Swift, and even Java) use the keyword `static` to mark variables with static storage duration.

Finally, and most importantly, I think "it implies storage lifetime" is intended to be this paragraph's thesis, but the phrasing seems to present it as a premise, and I'm not even entirely clear on what it means: in "X implies Y", normally X and Y refer to truth claims, but I can't tell what truth claims "it" or "storage lifetime" refer to. Even if I understood what this is saying, I'm not sure I'd agree that this paragraph makes the case for it. I'd suggest instead focusing this paragraph on a narrower but crisper thesis: using any spelling other than `static` here would be a surprising divergence from the languages that our users are most likely to be familiar with.

Author: Agree about inverting the structure a bit as the peer languages use this. I've also stolen some of your wording, as I tihnk that captures it better.

I wasn't trying to make the argument that the term's origins control how it is used today, I was trying to argue that the term's origins are _relevant_ to how we choose to use it. I don't think this is a fallacy. Sorry if it was unclear though, and since I don't think this is the most interesting point, I've largely removed it.

I've also removed the use of "implies" -- it has a very specific and formal meaning to you that wasn't my intent anyways. I meant something closer to "is associated with".

Reviewer: > I wasn't trying to make the argument that the term's origins control how it is used today, I was trying to argue that the term's origins are relevant to how we choose to use it. I don't think this is a fallacy.

Yeah, I agree that it's not a fallacy when it's used as a supporting argument or a tiebreaker, so I'd be fine with retaining it if you prefer. I just think it carries relatively little weight, so it shouldn't be what you lead with.
